src/Time/Tags.hpp

```diff
@@ -45,13 +45,21 @@ struct TimeStep : db::SimpleTag {
 
 /// \ingroup DataBoxTagsGroup
 /// \ingroup TimeGroup
-/// \brief Tag for compute item for current ::Time (from TimeId)
-struct Time : db::ComputeTag {
-  static std::string name() noexcept { return "Time"; }
+/// \brief Tag for compute item for ::Time of the current substep (from TimeId)
+struct SubstepTime : db::ComputeTag {
+  static std::string name() noexcept { return "SubstepTime"; }
   static auto function(const ::TimeId& id) noexcept { return id.time(); }
   using argument_tags = tmpl::list<TimeId>;
 };
 
+/// \ingroup DataBoxTagsGroup
+/// \ingroup TimeGroup
+/// \brief Tag for the current time as a double
+struct Time : db::SimpleTag {
```

Comment:
Can you make this a compute item, or add a compute item based on this? Then you don’t need to manually update this in `AdvanceTime`.

Reply:
I could, but I would just have to change it back when dense output lands and this can be the observation (or whatever) time.

Reply:
I feel like it's confusing to have a `Tags::Time` and a `Tags::SubstepTime` in the DataBox. It suggests the `Time` is different to the `SubstepTime` by not including the substep or something. Also, `Tags::Time` not being a compute item introduces the source of error that it is not updated correctly and not consistent with `time_id.substep_time()` anymore. These are concerns I have about the current state of the code, without knowing your plans for implementing dense output, which possibly fix these issues.

Reply:
@wthrowe I don't know if you saw the above comment as it didn't show up in the conversation...

Reply:
No, I missed this, sorry.

It may be confusing to have two times floating around, but there just *are* two quantities that are in some sense the current time.  I could get rid of the tag for the substep time and make consumers fetch it from the TimeId themselves.  I'm not sure if that's clearer.  It might be.  Thoughts?

I'm not sure what you are asking me to do with the second half of your comment.

Reply:
I think the following two options would work:

- Have three tags: `Tags::TimeId`, `Tags::Time` and `Tags::TimeValue`. I think that's what we had some time ago.
- Eliminate either `Tags::Time` or `Tags::TimeValue`. The name of the remaining tag should probably still follow its type, i.e. `Tags::Time` if it holds a `::Time` and `Tags::TimeValue` if it holds a `double`.

Now regarding compute items and dense output:

I think both `Tags::Time` and `Tags::TimeValue` should really be compute items. I don't know why your dense output implementation prevents that, and can't make an informed suggestion since I don't have code for dense output to look at. Perhaps dense output can use a prefixed `Tags::DenseOutput<Tags::Time>` or something.

Reply:
The problem with compute items doesn't have anything to do with the particular implementation or the code at all.  Dense output is evaluating things at a time that is not the same as the time the time stepper has evolved to.  There is no way to compute one of those times from the other because there is no particular relation between them.

Perhaps a pencil-and-paper-style example will make things clearer.  Let's integrate y'=y with y(0)=1 to e, so we can calculate e^e.  I don't want to type too much, so I'll use a step size of 1 and Euler's method.  Here are the states the calculation passes through:

|             | time-stepper time | dense-output time     | value   |
| ----------- | ----------------- | --------------------- | ------- |
| start       | 0                 | 0.0 (for convenience) | 1       |
| step 1      | 1                 | 1.0 (for convenience) | 2       |
| step 2      | 2                 | 2.0 (for convenience) | 4       |
| measurement | 2                 | e                     | 4 e - 4 |

The only way you can get the dense output time at the end is that it was part of the problem specification.  There is no way to compute it from the time-stepper time or any other quantity in the simulation.  When we are not performing a measurement it is convenient to set the dense-output time to the (value of the) time-stepper time so that you can use the same compute items for evolution and observing.

--------

I dislike using a prefix tag for dense output quantities because it makes doing the wrong thing (doing calculations on the substep time) easier than doing the right thing (doing calculations on the dense-output time).  So one thing I will insist on here is that if we end up with a tag called "Time" it *must* represent the dense-output time.

I would be OK with bringing back the old TimeValue tag in some form, but it didn't seem to be very useful in practice.  I think there was one thing that requested it and nothing ever added it to the DataBoxes in the main executables.

Reply:
Thanks for spelling out the example. I now understand that at the measurement step the time-stepper time and the dense-output time are _intentionally_ not the same. To make the distinction clearer, how about we rename `Tags::TimeId` -> `Tags::TimeStepId` (and also rename the class).

Then we can have a `Tags::Time` with type `double` that represent the time that should be used by everything to evaluate quantities. Hopefully that makes it clear that the `TimeStepId` represents the state of the time stepper, and one should not use it directly to compute physical quantities. We should also make notes on that in the documentation of both tags.

One confusion remains: There is a class `::Time` that is related to the time stepper, which one would wrongly expect to be the type of the `Tags::Time`. I would prefer to rename that class to something time-stepper-related, but can't think of a good suggestion.

Reply:
`TimeId -> TimeStepId` (and the tag): That is a much better name.  I will change that.

I am open to renaming `Time`, but I am also failing to come up with a better name.

Reply:
@wthrowe could you make this change so that we can get this PR in? would be useful to have this in the repository I think